aptos-move/block-executor/src/scheduler.rs

```diff
@@ -262,20 +295,48 @@ impl Scheduler {
             execution_idx: AtomicU32::new(0),
             validation_idx: AtomicU64::new(0),
             done_marker: CachePadded::new(AtomicBool::new(false)),
+            queueing_commits_lock: CachePadded::new(ArmedLock::new()),
+            commit_queue: ConcurrentQueue::<u32>::bounded(num_txns as usize),
         }
     }
 
     pub fn num_txns(&self) -> TxnIndex {
         self.num_txns
     }
 
+    pub fn add_to_commit_queue(&self, txn_idx: u32) {
+        self.commit_queue
+            .push(txn_idx)
+            .expect("Pushing to the commit_queue should never fail");
+    }
+
+    pub fn pop_from_commit_queue(&self) -> Result<u32, PopError> {
+        self.commit_queue.pop()
+    }
+
+    pub fn queueing_commits_mark_done(&self) {
+        self.queueing_commits_lock.unlock()
+    }
+
+    pub fn queueing_commits_arm(&self) {
+        self.queueing_commits_lock.arm()
+    }
+
+    pub fn should_coordinate_commits(&self) -> bool {
+        self.queueing_commits_lock.try_lock()
+    }
+
     /// If successful, returns Some(TxnIndex), the index of committed transaction.
     /// The current implementation has one dedicated thread to try_commit.
     /// Should not be called after the last transaction is committed.
     pub fn try_commit(&self) -> Option<TxnIndex> {
         let mut commit_state_mutex = self.commit_state.lock();
         let commit_state = commit_state_mutex.deref_mut();
-        let (commit_idx, commit_wave) = (&mut commit_state.0, &mut commit_state.1);
+        let (commit_idx, commit_wave) = commit_state;
+
+        if *commit_idx == self.num_txns {
```

Comment:
The comment says "should not be called after the last transaction is committed".
What change required doing this?